src/main/java/io/github/oliviercailloux/y2018/apartments/apartment/ApartmentFactory.java

```diff
@@ -154,39 +152,53 @@ private static double simulateRandomDraw(double mean, double deviation) {
 	/**
 	 * Call an API which generates a random address. This function aims at getting
 	 * the random address generated.
-	 *
-	 * @param retry indicates how many time the function has been called (Retry
-	 *              System). It will allow us to try escaping problems while
-	 *              contacting API (if it fails once, we try to call it again until
-	 *              the number specified in the class attribute NB_MAX_RETRY).
+	 * <p>
+	 * So, we generate a random latitude and a longitude and try to
+	 * {@link io.github.oliviercailloux.y2018.apartments.utils.JsonConvert#getAddressFromJson(String)
+	 * retrieve an address}
+	 * <p>
+	 * 
 	 * @return the address generated.
-	 * @throws IOException if we cannot contact the API generator.
+	 * @throws InvalidObjectException in case the API doesn't return a good format
+	 *                                after a certain number of attempts (RETRY)
 	 */
-	private static String getRandomAddress(int retry) throws IOException {
-
-		String address = "";
-		// Code from
-		// https://www.developpez.net/forums/d1354479/java/general-java/recuperer-reponse-d-adresse-http/
-		try (InputStream is = new URL(URL_API_ADDRESS).openConnection().getInputStream()) {
-			BufferedReader reader = new BufferedReader(new InputStreamReader(is, "utf-8"));
-			StringBuilder builder = new StringBuilder();
-			for (String line = reader.readLine(); line != null; line = reader.readLine()) {
-				builder.append(line + "\n");
+	private static String getOnlineRandomAddress() throws InvalidObjectException {
+		/**
+		 * Maximum test value in case the API does not return the
+		 * features.properties.labels field In the case of an API error (for example
+		 * 500, or others), jax-rs throws an error of type
+		 * <code>ClientErrorException</code>
+		 */
+		final int RETRY = 5;
+		// Latitude and longitude for Ile de France
+		final double LAT_IDF = 48.8_499_198d;
+		final double LONG_IDF = 2.6_370_411d;
+		// Open a new client JAX-RS
+		final Client client = ClientBuilder.newClient();
+		for (int i = 1; i <= RETRY; i++) {
+			// Latitude and longitude generation
+			double lat = Math.round(LAT_IDF * 10.0d) / 10.0d;
+			double lng = Math.round(LONG_IDF * 10.0d) / 10.0d;
+			String longitude = String.valueOf(lng) + String.valueOf(rand.nextInt((99_999 - 10000) + 1) + 10_000);
+			String latitude = String.valueOf(lat) + String.valueOf(rand.nextInt((99_999 - 10000) + 1) + 10_000);
+			// Call API
+			WebTarget target = client.target(URL_API_ADDRESS).queryParam("lon", longitude).queryParam("lat", latitude);
+			LOGGER.info("Address API Call : {}", target.getUri().toString());
+			String result = target.request(MediaType.TEXT_PLAIN).get(String.class);
+			try {
+				result = JsonConvert.getAddressFromJson(result);
+				client.close();
+				return result;
+			} catch (InvalidObjectException e) {
```

Comment:
Il est très inélégant de s’attendre à ce que la méthode `getAddressFromJson` renvoie `InvalidObjectException` pour signaler un problème de l’API de récupération d’adresse. Si c’est une possibilité réelle, il faut prévoir une façon de le tester qui affiche clairement qu’elle teste ceci.

Reply:
Fix: we created `AddressApiException` which extends `Exception`.